### PR TITLE
plugin/timeouts - Allow ability to configure listening server timeouts

### PR DESCRIPTION
**Pull request by rlees85**
### 1. Why is this pull request needed and what does it do?

Creates a "timeouts" plugin for configuring the listening server timeouts. Applies to TCP, DNS-over-TLS, DNS-over-HTTPS.

CoreDNS currently has a hard-coded 2 second read timeout for TCP/TLS connections and 5 seconds for HTTPS. This appears to be optimised for local network and/or Kubernetes environments where CoreDNS is most often used. Where CoreDNS is acting as a resolver on the public Internet this 2 second timeout can cause problems particularly on 3G/4G/5G connections where there might be lag.

It also removes the limit of queries per connection since we are enforcing timeouts this makes no sense and can cause some devices to think the DNS server is faulty.

It also removes the 128 requests per connection limit.

`Corefile` is also added to `.gitignore` for easier testing. Happy to revert this if preferred.

See more details in below.

### 2. Which issues (if any) are related?

https://github.com/coredns/coredns/issues/5780

### 3. Which documentation changes (if any) need to be made?

Plugin documentation - included.

### 4. Does this introduce a backward incompatible change or deprecation?

The only possible issue would be clients that *expect* the server to close the connection after 128 requests which feels very unlikely.

### CODE REVIEW

**Comment by rlees85** [comment]
I think this is as far as I can take this PR. Any guidance in making it good enough for merge would be appreciated. I am not 100% familiar with everything that's going on in CoreDNS and this is my first attempt at Go so only happy to take criticism/suggestions.

File: core/dnsserver/server.go [hunk 1]
```diff
@@ -81,6 +87,18 @@ func NewServer(addr string, group []*Config) (*Server, error) {
 		// append the config to the zone's configs
 		s.zones[site.Zone] = append(s.zones[site.Zone], site)
 
+		// set timeouts
+		for key, timeout := range site.Timeouts {
+			switch key {
+			case "idle":
```
Comment (Tantalor93) [inline core/dnsserver/server.go]: I am not a fan of these strings as keys, I would expect some static enums here, something like 
```
		for key, timeout := range site.Timeouts {
			switch key {
			case IdleTimeout:
				s.idleTimeout = timeout
			case ReadTimeout:
				s.readTimeout = timeout
			case WriteTimeout:
				s.writeTimeout = timeout
			}
		}
```
see other comments for config declaration

File: core/dnsserver/config.go [hunk 2]
```diff
@@ -53,6 +54,9 @@ type Config struct {
 	// TLSConfig when listening for encrypted connections (gRPC, DNS-over-TLS).
 	TLSConfig *tls.Config
 
+	// Timeouts for TCP, TLS and HTTPS servers.
+	Timeouts map[string]time.Duration
```
Comment (Tantalor93) [inline core/dnsserver/config.go]:
```suggestion
	Timeouts map[TimeoutType]time.Duration
```

and create new enum type
```
type TimeoutType uint8

const (
	IdleTimeout TimeoutType = iota
	ReadTimeout
	WriteTimeout
)
```

File: plugin/timeouts/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+Additionally some routers hold open connections when using DNS over TLS or DNS over HTTPS. Allowing
+a longer idle timeout helps performance and reduces issues with such routers.
+
+The *timeouts* "plugin" allows you to configure CoreDNS server read, write and idle timeouts.
+
+## Syntax
+
+~~~ txt
+timeouts READ [WRITE] [IDLE]
```
Comment (Tantalor93) [inline plugin/timeouts/README.md]: I would choose different approach to passing timeout durations, your approach forces me to provide all configurations when I want to override just IDLE timeout, so I would go with extra knobs configuration

```suggestion
timeouts {
	read DURATION
	write DURATION
	idle DURATION
}
```
=> If I want to override just write timeout I can do
```
timeouts {
	write 5s
}
```
if I want to override just idle timeout I can do 
```
timeouts {
	idle 5s
}
```

File: plugin/timeouts/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+## Examples
+
+Start a DNS-over-TLS server that picks up incoming DNS-over-TLS queries on port 5553 and uses the
+nameservers defined in `/etc/resolv.conf` to resolve the query. This proxy path uses plain old DNS.
+A 10 second read timeout, 20 second write timeout and a 60 second idle timeout have been configured.
+
+~~~
+tls://.:5553 {
+	tls cert.pem key.pem ca.pem
+	timeouts 10 20 60
```
Comment (Tantalor93) [inline plugin/timeouts/README.md]: also I would expect these arguments to accept format in GO duration, meaning  `5s`, `500ms`, etc

**Review by Tantalor93 — CHANGES_REQUESTED** [review]
Hello, thanks for your contribution! Although I have some objections regarding plugin configuration, please consider adjusting how the plugin will be configured

File: core/dnsserver/config.go (hunk 2 above)
Comment (chrisohaver) [reply core/dnsserver/config.go]: Could instead just be 3 separate properties in the struct. No need that I see for them to be in a map.

File: plugin/timeouts/README.md (hunk 3 above)
Comment (rlees85) [reply plugin/timeouts/README.md]: This makes a lot more sense. I am not entirely sure why I didn't do this to start with.

File: plugin/timeouts/README.md (hunk 4 above)
Comment (rlees85) [reply plugin/timeouts/README.md]: The revision I am going to submit will accept Go duration but if a plain number is specified it will assume that seconds was meant.

File: core/dnsserver/server.go [hunk 5]
```diff
@@ -81,6 +87,17 @@ func NewServer(addr string, group []*Config) (*Server, error) {
...
+			s.readTimeout = site.ReadTimeout
+		}
+		if site.WriteTimeout != time.Duration(0) {
+			s.writeTimeout = site.WriteTimeout
+		}
+		if site.IdleTimeout != time.Duration(0) {
+			s.idleTimeout = site.IdleTimeout
+		}
+
 		// copy tsig secrets
```
Comment (rlees85) [inline core/dnsserver/server.go:101]: Wondering if there is a better way to do this (multiple if)

File: core/dnsserver/server.go (hunk 1 above)
Comment (rlees85) [reply core/dnsserver/server.go]: I think this is resolved by no longer using a map. Please let me know if not

**Comment by rlees85** [comment]
Changes submitted for how the plugin should be configured

File: core/dnsserver/server_tls.go [hunk 6]
```diff
@@ -87,3 +97,7 @@ func (s *ServerTLS) OnStartupComplete() {
 		fmt.Print(out)
 	}
 }
+
+const (
+	tlsMaxQueries = -1
```
Comment (chrisohaver) [inline core/dnsserver/server_tls.go:102]: This change doesn't seem related to timeout configuration, and appears to be changing the default value from 0 to -1.

File: core/dnsserver/server.go (hunk 5 above)
Comment (chrisohaver) [reply core/dnsserver/server.go:101]: Maybe?  But it looks perfectly fine and easy to read as is.

File: core/dnsserver/server_tls.go (hunk 6 above)
Comment (rlees85) [reply core/dnsserver/server_tls.go:102]: You are right I think I didn't mention it in the pull request, but did in the issue its linked to, so my bad. The default was never set originally by Coredns so is defaulted to 128 here: https://github.com/miekg/dns/blob/master/server.go#18

The change is to remove the limit since we are having timeouts for idle connections anyway and an active connection has no good reason to limit the amount of requests it can make. Especially given a malicious actor could just keep reconnecting and sending another 128 anyway.

I'm also sure I found a post on Github somewhere where miekg mentioned he wasn't sure the point of the limit especially with timeouts set hence why I hard-coded to -1 (no limit) rather than trying to make it configurable.

Comment (chrisohaver) [reply core/dnsserver/server_tls.go:102]: > I think I didn't mention it in the pull request ...

You did mention it in the PR description, I just didn't see it before.  Thanks!

From the PR description...
> It also removes the limit of queries per connection since we are enforcing timeouts this makes no sense and can cause some devices to think the DNS server is faulty.

File: plugin/pkg/timeouts/timeouts.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"fmt"
+	"strconv"
+	"time"
+)
+
+// NewTimeoutFromArg returns a time.Duration from a configuration argument
+// (string) which has come from the Corefile. The argument has some basic
+// validation applied before returning a time.Duration.
+func NewTimeoutFromArg(arg string) (time.Duration, error) {
```
Comment (Tantalor93) [inline plugin/pkg/timeouts/timeouts.go]: why do we want this externalized? can we rather have it only as part of `plugin/timeouts` package?

File: plugin/pkg/timeouts/timeouts.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+	_, err := strconv.Atoi(arg)
+	if err == nil {
+		// If no time unit is specified default to seconds rather than
+		// GO's default of nanoseconds.
+		arg = arg + "s"
+	}
+
+	d, err := time.ParseDuration(arg)
+	if err != nil {
+		return time.Duration(0), fmt.Errorf("failed to parse timeout duration '%s'", arg)
```
Comment (Tantalor93) [inline plugin/pkg/timeouts/timeouts.go]: just a nitpick, unnecessary boxing of 0 to time.Duration (I think there are multiple occurrences of this unnecessary boxing)
```suggestion
		return 0, fmt.Errorf("failed to parse timeout duration '%s'", arg)
```

File: plugin/timeouts/timeouts.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+		for c.NextBlock() {
+			block := c.Val()
+			timeoutArgs := c.RemainingArgs()
+			if len(timeoutArgs) != 1 {
+				return c.ArgErr()
+			}
+
+			timeout, err := timeouts.NewTimeoutFromArg(timeoutArgs[0])
+			if err != nil {
+				return err
```
Comment (Tantalor93) [inline plugin/timeouts/timeouts.go]:
```suggestion
		return c.Err(err.Error())
```
to also propagate the line number of Corefile to the error message

File: plugin/pkg/timeouts/timeouts_test.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
+package timeouts
+
+import (
+	"testing"
+	"time"
+)
+
+func TestNewTimeoutFromArg(t *testing.T) {
```
Comment (Tantalor93) [inline plugin/pkg/timeouts/timeouts_test.go]: I think this would be better as table test
```
func TestNewTimeoutFromArg(t *testing.T) {
	tests := []struct {
		name    string
		arg     string
		wantErr bool
		want    time.Duration
	}{
		{
			name: "valid GO duration - seconds",
			arg:  "30s",
			want: 30 * time.Second,
		},
		{
			name: "valid GO duration - minutes",
			arg:  "2m",
			want: 2 * time.Minute,
		},
		{
			name: "number - fallback to seconds",
			arg:  "30",
			want: 30 * time.Second,
		},
		{
			name:    "invalid timeout",
			arg:     "twenty seconds",
			wantErr: true,
		},
		{
			name:    "timeout is too low",
			arg:     "0",
			wantErr: true,
		},
		{
			name:    "timeout is too high",
			arg:     "24h1s",
			wantErr: true,
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			actual, err := NewTimeoutFromArg(test.arg)
			if test.wantErr && err == nil {
				t.Error("error was expected")
			}
			if !test.wantErr && err != nil {
				t.Error("error was not expected")
			}

			if test.want != actual {
				t.Errorf("expected '%v' got '%v'", test.want, actual)
			}
		})
	}
}
```

**Review by Tantalor93 — CHANGES_REQUESTED** [review]
overall looks good to me, I would just make few quality improvements 👍

File: plugin/pkg/timeouts/timeouts.go (hunk 7 above)
Comment (Tantalor93) [reply plugin/pkg/timeouts/timeouts.go]: if we want this externalized I think we should try to think of better name of a package and function, this is quite common duration parsing, not necessarily timeout
```suggestion
package durations

func NewDurationFromArg(arg string) (time.Duration, error) {
```

Comment (rlees85) [reply plugin/pkg/timeouts/timeouts.go]: I have no good reason to externalise it. Looking at the TLS plugin it looked like anything beyond very basic parsing was externalised. I would be happy to try and bring it together if it was felt it would be better.

**Comment by rlees85** [comment]
Thanks @Tantalor93 I will try and get on these improvements when I get a moment

File: plugin/pkg/timeouts/timeouts.go (hunk 7 above)
Comment (rlees85) [reply plugin/pkg/timeouts/timeouts.go]: I think I would leave it (mostly) externalised but rename to durations. It would do the basic parsing and default durations passed as integers to seconds as it does currently. This might be useful in other areas of the app when parsing durations rather than having to duplicate code.

The check for "max" and "min" durations will be moved into the `plugin/timeouts` bit since durations parsed in other parts of the code may not have such constraints.

Would you be happy with that approach? i will convert the tests to table tests as well

Comment (Tantalor93) [reply plugin/pkg/timeouts/timeouts.go]: Sounds good 👍

**Comment by rlees85** [comment]
Hopefully that should cover the comments unless I've messed something up (i've ran go tests/etc locally so hopefully not!)

**Review by Tantalor93 — APPROVED** [review]
LGTM, thanks for your contribution!
